crates/term_input_macros/src/lib.rs

```diff
@@ -13,10 +17,31 @@ pub fn byte_seq_parser(input: TokenStream) -> TokenStream {
         fn_return_type,
         rules,
     } = syn::parse_macro_input!(input as syntax::Input);
-    let fn_body = tree::build_decision_tree(rules);
 
-    quote!(fn #fn_name(buf: &[u8]) -> Option<(#fn_return_type, usize)> {
+    let is_valid_key = build_is_valid_key(&rules);
+
+    let fn_body = tree::build_decision_tree(rules);
+    let byte_parser = quote!(fn #fn_name(buf: &[u8]) -> Option<(#fn_return_type, usize)> {
         #fn_body
-    })
-    .into()
+    });
+
+    quote!(#is_valid_key #byte_parser).into()
+}
+
+fn build_is_valid_key(rules: &[Rule]) -> TokenStream2 {
+    let keys = rules
+        .iter()
+        .map(|r| &r.value.0)
+        .collect::<HashSet<_>>()
+        .into_iter();
+
+    quote!(
+        pub fn is_valid_key(k: Key) -> bool {
+            match k {
+                #(#keys => true,)*
+                Key::AltChar(_) | Key::Char(_) => true,
```

Comment:
This case is specific to our use of the macro in `term_input`, so I think it's better to handle it in a function in that module. We could have a function in `term_input` that checks for these cases, then for the default case calls the macro-generated function.